File: SherlockAndAnagrams.py
#!/usr/local/bin/python3

import sys

def checkAnagram(s1, s2):
    mem = [0] * 26
    for c in s1:
        mem[ord(c) - ord('a')] += 1
    for c in s2:
        mem[ord(c) - ord('a')] -= 1
    for i in mem:
        if i != 0:
            return False
    return True

def sherlockAndAnagrams(s):
    result = 0
    for length in range(1, len(s)):
        for start in range(0, len(s) - length):
            for end in range(start + 1, len(s) - length + 1):
                if checkAnagram(s[start:(start+length)], s[end:(end+length)]):
                    result += 1
    return result

q = int(input().strip())
for a0 in range(q):
    s = input().strip()
    result = sherlockAndAnagrams(s)
    print(result)

File: SherlockAndAnagrams.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

bool checkAnagram(const string& str, int start, int end, int count) {
    vector<int> mem(26, 0);

    for (int i = 0; i < count; i++) {
        mem[str[start+i] - 'a'] += 1;
        mem[str[end+i] - 'a'] -= 1;
    }

    for (auto i : mem) {
        if (i != 0) {
            return false;
        }
    }

    return true;
}

int sherlockAndAnagrams(const string& s) {
    int result = 0;
    for (int length = 1; length < s.length(); length++) {
        for (int start = 0; start < s.length() - length; start++) {
            for (int end = start + 1; end < s.length() - length + 1; end++) {
                if (checkAnagram(s, start, end, length)) {
                    result += 1;
                }
            }
        }
    }

    return result;
}

int main(void) {
    int n;
    cin >> n;
    string s;
    for (int i = 0; i < n; i++) {
        cin >> s;
        cout << sherlockAndAnagrams(s) << endl;
    }
}
